python/paddle/distributed/auto_parallel/intermediate/pipeline_parallel.py

```diff
@@ -85,11 +86,38 @@ def forward_pre_hook(layer, input):
             # TODO(deepllz): support in the future
             return input
 
+        # step1: set every layer's own pipeline_stage_index
         split_layer_names = list(self.split_spec.keys())
-        for index, name in enumerate(split_layer_names):
+        sublayer_names = [name for name, _ in model.named_sublayers()]
+        # Mark which layer is the next pipeline stage
+        pipline_layer_mark = [0 for _ in range(len(sublayer_names))]
+        for split_layer_name in split_layer_names:
+            split_point = self.split_spec[split_layer_name]
+            index = sublayer_names.index(split_layer_name)
+            if split_point == SplitPoint.END:
+                is_valid = False
+                for i in range(index + 1, len(sublayer_names)):
+                    if not sublayer_names[i].startswith(split_layer_name):
+                        pipline_layer_mark[i] = 1
+                        is_valid = True
+                        break
+                assert (
+                    is_valid
+                ), f"the last layer:{split_layer_name} must not be SplitPoint.END, please check the split_spec"
+            else:
+                raise NotImplementedError(
+                    "SplitPoint.BEGINNING is not supported currently"
+                )
+                pipline_layer_mark[index] = 1
```

Comment:
There is currently no problem, but we will support `SplitPoint.BEGIN` in the future, it may be necessary to check whether it has already been marked (if it has been marked, report an error?). For example, split_spec is written as follows: `split_spec={"layer.0" : SplitPoint.END, "layer.1" : SplitPoint.BEGIN}`, in first loop, layer.0 will be marked on layer.1, and then in second loop, it will be marked again by layer.1

Reply:
According to the current logic, there will be no problem with repeated markings, because the marking method is to assign a value of 1.